addons/l10n_it_edi/models/account_edi_format.py

```diff
@@ -406,17 +411,27 @@ def _import_fattura_pa(self, tree, invoice):
                 invoice = self.env['account.move']
             first_run = False
 
+            elements_buyer_vat = tree.xpath('//CessionarioCommittente//IdCodice')
+            elements_buyer_codice_fiscale = tree.xpath('//CessionarioCommittente//CodiceFiscale')
+
+            elements_seller_vat = tree.xpath('//CedentePrestatore//IdCodice')
+            elements_seller_codice_fiscale = tree.xpath('//CedentePrestatore//CodiceFiscale')
+
             # Type must be present in the context to get the right behavior of the _default_journal method (account.move).
             # journal_id must be present in the context to get the right behavior of the _default_account method (account.move.line).
-            elements = tree.xpath('//CessionarioCommittente//IdCodice')
-            company = elements and self.env['res.company'].search([('vat', 'ilike', elements[0].text)], limit=1)
-            if not company:
-                elements = tree.xpath('//CessionarioCommittente//CodiceFiscale')
-                company = elements and self.env['res.company'].search([('l10n_it_codice_fiscale', 'ilike', elements[0].text)], limit=1)
-                if not company:
-                    # Only invoices with a correct VAT or Codice Fiscale can be imported
-                    _logger.warning('No company found with VAT or Codice Fiscale like %r.', elements[0].text)
-                    continue
+
+            buyer_company = _search_company(vat_elements=elements_buyer_vat,
+                                            codice_fiscale_elements=elements_buyer_codice_fiscale)
+            seller_company = _search_company(vat_elements=elements_seller_vat,
+                                             codice_fiscale_elements=elements_seller_codice_fiscale)
+
+            if not (buyer_company or seller_company):
+                # Only invoices with a correct VAT or Codice Fiscale can be imported
+                _logger.warning('No company found with VAT or Codice Fiscale from the buyer or seller section.')
+                continue
+
+            company = buyer_company or seller_company
+            incoming = company == buyer_company
```

Comment:
```suggestion
              for incoming, vat_xpath, codice_fiscale_xpath in (
                  (True, '//CessionarioCommittente//IdCodice', '//CessionarioCommittente//CodiceFiscale'),
                  (False, '//CedentePrestatore//IdCodice', '//CedentePrestatore//CodiceFiscale'),
              ):
                  company = _search_company(vat_elements=tree.xpath(vat_xpath), codice_fiscale_elements=tree.xpath(codice_fiscale_xpath))
                  if company:
                      break
              else:
                  _logger.warning('Cannot find any company that matches the VAT or Codice Fiscale found in the buyer and seller sections.')
                  continue
```

`incoming` is valid also outside the `for` loop.

Reply:
If we are importing in a `sale` journal maybe we could have sth like:

```py
            buyer_or_seller_mapping = [
                (True, '//CessionarioCommittente//IdCodice', '//CessionarioCommittente//CodiceFiscale'),
                (False, '//CedentePrestatore//IdCodice', '//CedentePrestatore//CodiceFiscale'),
            ]
            if <coming_from_sale_journal>:
                mapping = mapping[::-1]
            for incoming, vat_xpath, codice_fiscale_xpath in buyer_or_seller_mapping:
```

Reply:
In the new version I basically followed your suggestion. Except 
- I made the cases a bit more explicit
- used a dictionary instead of a list (since we need the items also to parse the partner)
I also restricted the upload to the allowed companies (to restrict the upload button)